### PR TITLE
feat: add FSST string compression

### PR DESCRIPTION
**Pull request by broccoliSpicy**
#2415 
This is a naive implementation of the [FSST](https://www.vldb.org/pvldb/vol13/p2649-boncz.pdf), it is fairly tested so that it can be used in correctness proof for future optimization, and a mark for future performance engineering.

this is not complete, but enough for one git commit, it's just, I am going through something and a bit debilitated, I hope pushing a PR and make the progress live may motivate me more.

Question 1: why don't use the original C++ implementation in the paper and do some Rust/C++ FFI?
    the original C++ implementation is not tuned for our workload
    the original C++ implementation doesn't work directly with arrow string_array interface
    
Question 2: how to use this FSST implementation?
this is the designed interface, 
```rust
pub fn fsst_compress(_input_buf: &[u8], _input_offsets_buf: &[i32], _output_buf: &mut Vec<u8>, _output_offsets_buf: &mut Vec<i32>) -> Result<()> {
    Ok(())
}

pub fn fsst_decompress(_input_buf: &[u8], _input_offsets_buf: &[i32], _output_buf: &mut Vec<u8>, _output_offsets_buf: &mut Vec<i32>) -> Result<()> {
    Ok(())
}

```
TODO: 
1: more comments and documentations
2: benchmark this implementation(how can I get test data for our intended workload?)
3: push string encoding from lance logical encoding to lance physical encoding
4: lance string compression trait in lance physical encoding
5: implement FSST for lance string compression trait
6: implement SIMD optimization in the paper
7: maybe we could make FSST as a separate rust crate
8: predicate push down to FSST decompression

a few questions:
1: does lance intended to support big-endian machine?
2: what kind of string dataset is preferable for our use?

### CODE REVIEW

**Comment by westonpace** [comment]
This is very cool!

I'll look through this in more detail tomorrow morning on Monday morning.

> this is not complete, but enough for one git commit, it's just, I am going through something and a bit debilitated, I hope pushing a PR and make the progress live may motivate me more.

I think this PR is plenty for a single commit.  We don't need to tackle it all in one go, having just the compression routines is a good start.  We can help out with some of the integration into the larger system too.

> Question 2: how to use this FSST implementation?
> this is the designed interface,

Is there no need for a symbol table?  Or is the symbol table somehow encoded into the output buffer?

>  2: benchmark this implementation(how can I get test data for our intended workload?)

I think a compelling test set would be one of the training datasets used to create ML semantic search models.  For example, the [MS Marco](https://microsoft.github.io/msmarco/) dataset is very large (probably just need a subset of it for benchmarking): https://msmarco.z22.web.core.windows.net/msmarcoranking/fulldocs.tsv.gz

These are English language strings that I think would be a good candidate for FSST.

>  3: push string encoding from lance logical encoding to lance physical encoding

There is an effort in progress to do this because we needed it for dictionary encoding too.  I agree it is needed: https://github.com/lancedb/lance/pull/2426

>  7: maybe we could make FSST as a separate rust crate

Sure, we'd be happy to host it on our org (lancedb) or use any other you have in mind.

>  8: predicate push down to FSST decompression

Predicate pushdown is partly wired up at the logical level now but I haven't really thought of a good way to push it down into the physical level.  One challenge is that I've been trying to avoid making datafusion a dependency of `lance-file` (it is large and the project benefits from fast build / debug cycles) so at the moment expressions, from the `lance-file` perspective, are just substrait encoded `Bytes`.  Some kind of lightweight typed expression library would be useful here.

**Comment by broccoliSpicy** [comment]
> Is there no need for a symbol table? Or is the symbol table somehow encoded into the output buffer?

the symbol table needs to be serialized/deserialized to the output buffer during compression/decompression, which I haven't implemented and forgot to mention, sorry for the confusion.

> Sure, we'd be happy to host it on our org (lancedb) or use any other you have in mind.

I will be so honored if it can be hosted on lancedb! I have a fair amount of work to do in optimization though.

> Predicate pushdown is partly wired up at the logical level now but I haven't really thought of a good way to push it down into the physical level. One challenge is that I've been trying to avoid making datafusion a dependency of lance-file (it is large and the project benefits from fast build / debug cycles) so at the moment expressions, from the lance-file perspective, are just substrait encoded Bytes. Some kind of lightweight typed expression library would be useful here.

I can only find duckDB has fsst compression, they use the original implementation in the paper and doesn't support predicate pushdown.
this might be a opportunity for us to get some state-of-art string processing benchmark results but I am not sure how good the result would be, if the benchmark improvement is not significant, we might be better off to not have this predicate push down

**Review by westonpace — COMMENTED** [review]
Ok, took a closer look.  This looks really good so far.

It looks like you have the basic structures (symbol, symbol table, hashing table) and the code to build the symbol table, and then also routines for encoding & decoding with the symbol table (`compress_bulk` and `decompress_bulk`).

You have some basic unit tests for the structures.
There are some commented out end-to-end tests.  Why are these commented out?

You added an end-to-end test based on the MS Marco dataset, this can probably evolve to a benchmark.

>  1: does lance intended to support big-endian machine?

I missed this question.  The answer, for the moment, is no.  We don't have any testing on big endian machines and so I think its reasonable to assume that some things might fail today.

Are you asking because this code will fail on a big endian machine (I see there are some `ne_bytes` routines)?  I think this is fine as long as we clearly document that fact.  Ideally we could also detect if we are in a big endian machine and panic.

---

How can we help?

I agree that making this a standalone crate will be useful.  I don't see any existing rust crates for FSST.  Probably the easiest thing to do initially would be to make this a standalone crate within the same lance mono-repo and then add code to publish the crate as part of the lance publish.

It looks like the only `lance` dependency at the moment is `lance_core::Result` and so we'd just need to create an `FsstResult` type which shouldn't be too much work.  This crate could then stand on its own.

Alternatively, we can start a new repo for encodings utilities or even a repo just for FSST.  I'm not sure (yet) what else would go in those repos and so we can wait and do that later I suppose.

This can all be part of future PRs too.  I think the only things we need to do at the moment to merge this PR are to clean up the formatting and figure out which commented bits to remove.

Once https://github.com/lancedb/lance/pull/2426 merges we can create a PR to connect this to the lance file format.  I can do this (given the compression & decompression routines I expect this to be pretty easy) or I'm happy to leave it to you if you want to learn more about the encoding & decoding traits.

**Comment by broccoliSpicy** [comment]
> Ok, took a closer look. This looks really good so far.

thank you so much for looking into these such primitive code, I feel pumped with these discussions and feedbacks!

the current code is more intended to set up the testing and benchmark framework so I can experiment with the algorithm, I expect the final version be very different

File: rust/lance-encoding/src/encodings/compression_algo/fsst.rs
Comment (broccoliSpicy) [reply rust/lance-encoding/src/encodings/compression_algo/fsst.rs:16]: the high 32 bit in a symbol.icl is not used, so I think a better way to represent a empty hash slot would be have a special high 32 bit, I get this `8` from the C++ implementation's comment
```C++
// in the hash table, the icl field contains (low-to-high) ignoredBits:16,code:12,length:4
#define FSST_ICL_FREE ((15<<28)|(((u32)FSST_CODE_MASK)<<16)) // high bits of icl (len=8,code=FSST_CODE_MASK) indicates free bucket
```
[C++ implemention](https://github.com/cwida/fsst/blob/50dd308befd5121bd1121428a833f42cce6e1f9d/libfsst.hpp#L139)

File: rust/lance-encoding/src/encodings/compression_algo/fsst.rs
Comment (broccoliSpicy) [reply rust/lance-encoding/src/encodings/compression_algo/fsst.rs]: this was a mistake, the fmt::Debug of Symbol should have a specified implementation

File: rust/lance-encoding/src/encodings/compression_algo/fsst.rs
Comment (broccoliSpicy) [reply rust/lance-encoding/src/encodings/compression_algo/fsst.rs:81]: yes, in the original C++ implementation, it is said to have performance benefits during compression when we store this value

File: rust/lance-encoding/src/encodings/compression_algo/fsst.rs
Comment (broccoliSpicy) [reply rust/lance-encoding/src/encodings/compression_algo/fsst.rs]: this is actually a mistake, the alphabetical order of a string with length <= 8 can not be acquire by comparing their value interpreted as u64, in both little-endian and big-endian machines.
I copied this logic from the original C++ implementation and overlooked it

**Comment by broccoliSpicy** [comment]
benchmark results track:
ebc8a4ef44bc4133fee4e3169a0f32bd53fa4d12
Screenshot 2024-06-20 at 7 49 27 PM

machine info:
`Linux 192 5.10.0-28-amd64 #1 SMP Debian 5.10.209-2 (2024-01-31) x86_64 GNU/Linux`

**Comment by broccoliSpicy** [comment]
a few more questions @westonpace:
I am playing with the algorithm with the input buffer size 16MB, assuming we spawn a encoding task when we have 16MB arrow string value data(not counting the offsets), what's the appropriate size here?

I have a x86 CPU with AVX512, and a macbook pro 2023 with Apple M2 Pro, I don't have access to other vector extension instruction sets. I think we should prioritize AVX512?

I couldn't get the compression ratio up to 1.6 for the third column of [MS Marco](https://microsoft.github.io/msmarco/), this is expected and in align with the original paper results, FSST achieves better compression ratio on small similar data sets(3.6 on TPCH c_name column for 9e1f0844ef82b1666553c5977f0df162bf691f38), is this good enough or should I invest more time on this?

**Comment by westonpace** [comment]
>  I am playing with the algorithm with the input buffer size 16MB, assuming we spawn a encoding task when we have 16MB arrow string value data(not counting the offsets), what's the appropriate size here?

The default right now is 8MiB so not far off.  It's configurable and will often be some value over 8MiB (e.g. if we have 7MB of data and a new array comes in with 2MB we will just make a 9MB page).

> I have a x86 CPU with AVX512, and a macbook pro 2023 with Apple M2 Pro, I don't have access to other vector extension instruction sets. I think we should prioritize AVX512?

That's up to you as this is your contribution.  I think it's fine to prioritize AVX512 as long as we panic if it isn't present (we have flags to do runtime checks).  Eventually, we will probably want at least an AVX2 version as a fallback (I'm pretty sure this is our minimum supported) if we want to make this encoding a default and not opt-in.

> is this good enough or should I invest more time on this(I am afraid the improvement would be marginal)?

I think it's fine.  No need to tackle everything all at once.  We can investigate other encodings for columns like that in the future.

**Comment by broccoliSpicy** [comment]
> The default right now is 8MiB so not far off. It's configurable and will often be some value over 8MiB (e.g. if we have 7MB of data and a new array comes in with 2MB we will just make a 9MB page).

just to make sure, when we spawn a encoding task with 8MiB input, the output page is much less, I assume this kind of segmentation is handled by our code?

**Comment by westonpace** [comment]
> just to make sure, when we spawn a encoding task with 8MiB input in memory, the output buffer in memory is much less, I assume this kind of segmentation is handled by our code?

At the moment, we always write the encoded buffers to a page, even if the output buffers are much smaller.  In the future we can add code to accumulate output buffers until 8MiB of encoded data has been built up.  The only tricky part is that, once we have chosen an encoding tree, we need to make sure we keep using that encoding tree until we flush a page.

We haven't needed it yet since we don't have any compressive encodings.

I think it's probably fine for now and we can optimize later.

**Comment by westonpace** [comment]
FYI, the string encoder has moved from a field encoder to an array encoder.  This should make it considerably easier to integrate.  I will make an integration PR to allow testing FSST end-to-end soon.

**Comment by broccoliSpicy** [comment]
> FYI, the string encoder has moved from a field encoder to an array encoder. This should make it considerably easier to integrate. I will make an integration PR to allow testing FSST end-to-end soon.

Thank you so much! 
excited to see the integration!
excited to see more and more encoding go into lance!

**Comment by westonpace** [comment]
This turned out to be pretty straightforward to integrate.  Here's a commit that integrates FSST: https://github.com/lancedb/lance/commit/71ac855c232cdf952cde4661d54e515d6978901d

With FSST integrated the unit tests passed (:tada:) and I was able to encode the TPC-H lineitems table.  However, I'm currently getting a panic on the decode.  I can investigate in more detail soon (wanted to get the integration done today and didn't dig too deep into FSST code).

That being said, we should get this PR merged, as is, and then figure out the TPC-H panic and do the integration in a follow-up PR.  Working in small chunks will be helpful.  So, priority:

* There are some tests that didn't compile (missing symbol from slice I think).  Let's add that method back in or remove those tests.
* Please clean up the commented blocks (feel free to remove commented out tests that require setup and we can add them back in later).
* Please address any warnings and CI failures

Once that is done we can merge this PR.

A few things popped up during the integration that we can then address in a follow-up:

* Lance encoders receive the arrays to encode as a slice of arrays.  This is because the column encoder collects arrays until it has enough data to justify encoding.  Currently the `fsst::fsst::compress` function expects a single array.  This means I need to concatenate the input arrays before calling `fsst::fsst::compress` which I'd like to avoid.  I think this just means adding a top-level `compress_many`.
* It would be very helpful if we can save the symbol table in a separate buffer and read the symbol table from a separate buffer instead of inserting it into the beginning of the bytes array.  This is because the symbol table is going to get stored somewhere else (the metadata) and I end up needing to do some extra copies to work around this.
* The `fsst::fsst::compress` routine panics if given very small inputs (e.g. not enough space for the symbol table).  I don't think this is really a problem because we should probably have a size limit for using FSST anyways (and that is what I ended up doing to avoid this).
* I need to change the core encoding traits to work with `Vec<u8>` instead of `BytesMut` I think.  This isn't just an FSST issue, I have encountered this issue elsewhere.  It's difficult to create a new `BytesMut` from existing data.  This introduced a few copies on the decode side.

**Comment by broccoliSpicy** [comment]
cool!
I will start doing it right away!

**Comment by broccoliSpicy** [comment]
> * It would be very helpful if we can save the symbol table in a separate buffer and read the symbol table from a separate buffer instead of inserting it into the beginning of the bytes array.  This is because the symbol table is going to get stored somewhere else (the metadata) and I end up needing to do some extra copies to work around this.

This is open to discussion, I don't think we need to store the symbol table somewhere else as metadata, the FSST symbol table itself doesn't offer any filter capability or statistics of a single string attribute. 

> The fsst::fsst::compress routine panics if given very small inputs (e.g. not enough space for the symbol table). I don't think this is really a problem because we should probably have a size limit for using FSST anyways (and that is what I ended up doing to avoid this).

the current code has this logic of store string as is when input is small, I set it to 0 during development

**Comment by westonpace** [comment]
> This is open to discussion, I don't think we need to store the symbol table somewhere else as metadata, the FSST symbol table itself doesn't offer any filter capability or statistics of a single string in this column, it is more about substrings in this column segment.

Fair, but it is small, and if we can cache it then that is one less bit of data we need to read when searching.  LanceDB does vector search and some of our customers are bound by "max 5500 read requests per file per second".  However, we can cache it even if stored in bytes buffer I suppose, that will help keep the metadata small.

**Comment by broccoliSpicy** [comment]
> With FSST integrated the unit tests passed (:tada:) and I was able to encode the TPC-H lineitems table (and saw some compression). However, I'm currently getting a panic on the decode. I can investigate in more detail soon (wanted to get the integration done today and didn't dig too deep into FSST code).

I actually don't know how to do end-to-end test for lance, could you give some reference so I can also investigate on this? @westonpace

**Comment by broccoliSpicy** [comment]
the current decompression panic issue does need more discussion:
there is an asymmetry in our ArrayEncoder/PageDecoder, for one output buffer that a encoding algorithm produces, we may spawn many decoding tasks, and these decoding task all tries to interpret decode meta data from these decode input buffers

**Comment by westonpace** [comment]
> there is an asymmetry in our ArrayEncoder/PageDecoder, for one output buffer that an encoding algorithm produces, we may spawn many decoding tasks, and these decoding tasks all tries to interpret decode meta data from these decode input buffers

Yes, this is correct.  We slice the buffers into batches (I normally use 8Ki and can't remember if the default is 1Ki or 8Ki).  We slice the offsets buffer into equal sized chunks and then grab the corresponding byte ranges.  Then we create a decode task for each one.  Since each chunk receives "whole strings only" (e.g. a string is not split across chunks) I think this works ok yes?  I don't remember if I cloned the symbol table for each decode task or not but we can easily wrap it in an Arc to prevent deep clones (decode should only need read-only access to symbol table)

**Comment by broccoliSpicy** [comment]
> Since each chunk receives "whole strings only" (e.g. a string is not split across chunks) I think this works ok yes?

the current fsst::decompress is designed in a way that assumes the input is a wholly compressed block and tries to interpret symbol table from the beginning of the buffer, I should think more about this

**Comment by westonpace** [comment]
> I actually don't know how to do end-to-end test for lance, could you give some reference so I can also investigate in this? @westonpace

If you merge my PR as it is applied you can install the python bindings by running the following.  `maturin develop` will install the wheel into your current conda / venv in editable mode.

```
cd python
maturin develop --release
```

The following python script is what I was using for testing:

```
from datetime import datetime
import glob
import os

import pyarrow as pa
import pyarrow.parquet as pq
import pyarrow.dataset as ds

import lance
from lance.file import LanceFileReader, LanceFileWriter

from lance.tracing import trace_to_chrome

trace_to_chrome(level="trace")

PATH = "/home/pace/dev/data/lineitem_10.parquet"

arr_dataset = ds.dataset(PATH)

schema = ds.dataset(PATH).schema

# for file in glob.glob("/home/pace/dev/data/lineitem_10.parquet.*"):
#     pq_size = os.path.getsize(file)
#     pq_size_gb = pq_size / 1024.0 / 1024.0 / 1024.0
#     print(f"Parquet file size: {pq_size_gb} GiB")

#     pq_metadata = pq.ParquetFile(file).metadata
#     num_row_groups = pq_metadata.num_row_groups
#     print(f"Parquet file num row groups: {num_row_groups}")

#     start = datetime.now()
#     arr_dataset = ds.dataset(PATH)
#     arr_dataset.to_table()
#     end = datetime.now()

#     duration = (end - start).total_seconds()
#     print(f"Parquet (pyarrow) Read Time (lineitem_10): {duration}s")

if not os.path.exists("/tmp/tpch.lancev2"):
    start = datetime.now()
    arr_dataset = ds.dataset(PATH)
    with LanceFileWriter("/tmp/tpch.lancev2", arr_dataset.schema) as writer:
        for batch in arr_dataset.to_batches():
            writer.write_batch(batch)
    end = datetime.now()
    duration = (end - start).total_seconds()
    print(f"V2 Write Time (lineitem_10): {duration}s")

lv2_size = os.path.getsize("/tmp/tpch.lancev2")
lv2_size_gb = lv2_size / 1024.0 / 1024.0 / 1024.0
print(f"Lance v2 file size: {lv2_size_gb} GiB")

start = datetime.now()
reader = LanceFileReader("/tmp/tpch.lancev2")
for batch in reader.read_all(batch_size=8 * 1024).to_batches():
    pass
end = datetime.now()
print(f"V2 Read Time (lineitem_10): {(end - start).total_seconds()}s")
```

The `lineitem_10.parquet` file can be downloaded [here](https://drive.google.com/file/d/1rA-tEDEA5nyahSozL9UhrpXiaxA2lYHn/view?usp=sharing).

**Comment by westonpace** [comment]
> I think we can at least think about the alternatives: instead of spawn decoding task based on row number batches, what else can we do?

This is how query engines operate.  They stream small CPU-cache sized batches of data through the system.  Any other approach generates expensive cache<->CPU copies which defeats any potential SIMD / compression win.

**Comment by westonpace** [comment]
> the thing is, we need to tell fsst which row number starts a new compressed block and instruct fsst to interpret symbol table at that buffer, this can get nasty,
we can have fsst output 3 buffers, one for compressed bytes, one for offsets, one for symbol table, but we still need to tell the PageDecoder which row number begins a new compression unit and needs to fetch new symbol table

Decode batches will:
* Never start in the middle of a string
* Never span multiple pages / chunks

Let's assume we are decoding and we have 6MB of compressed data.  We have 5001 offsets and so we have 5000 rows.  First, we decode the offsets.  Then we split the offsets into 5 chunks of 1000 rows.  Maybe the first chunk is 2MB of data, the second one is 512KB, and the next 2 are 1MB and the last one is 1.5MB.  This is all fine.  No string is split between chunks.  We should be able to decode each of them in parallel independently I think (they will all need read access to the symbol table but none of them modify it)

**Comment by broccoliSpicy** [comment]
I changed compression/decompression's interface to accept symbol table and merged @westonpace's integration code, but there are still many bugs

it might be a trouble for us when dealing with `page fragmentation after encoding`, we only have one page encoding info per page, also in page bridging

putting parallelism inside one decode task may mess our structure, I am a bit hesitate about this, another reason is that we may be able to achieve good compression ratio while keeping the decompression very very lightweight, by using only lightweight algorithms

**Comment by broccoliSpicy** [comment]
end-to-end test result using `ms marco`:


Screenshot 2024-07-01 at 6 18 46 PM


there is a offset overflow bug when using full `ms marco` dataset(22 GiB)

the batch seems not scheduled in a column major manner but a row major manner

to reproduce, run 
```python
from datetime import datetime
import os
import pyarrow.csv as csv
import pyarrow as pa

import lance
from lance.file import LanceFileReader, LanceFileWriter

from lance.tracing import trace_to_chrome

# Path to the TSV file
file_path = "/home/x/fulldocs_500_k.tsv"

# Read the TSV file into a PyArrow table
table = csv.read_csv(file_path, read_options=csv.ReadOptions(column_names=['column1', 'column2', 'column3'], block_size = 1024 * 1024 * 1024), parse_options=csv.ParseOptions(delimiter='\t'))

with LanceFileWriter("/tmp/fulldocs.lancev2", table.schema) as writer:
    writer.write_batch(table)

original_size = os.path.getsize(file_path)
original_size_mb = original_size / 1024.0 / 1024.0 
print(f"original file size: {original_size_mb} MiB")

lv2_size = os.path.getsize("/tmp/fulldocs.lancev2")
lv2_size_mb = lv2_size / 1024.0 / 1024.0 
print(f"Lance v2 file size: {lv2_size_mb} MiB")

print(f"compression ratio: {original_size / lv2_size}")
start = datetime.now()

reader = LanceFileReader("/tmp/fulldocs.lancev2")
table_read = reader.read_all().to_table()

assert table.equals(table_read), "The original table and the table read back from the Lance file are not identical."
```
in b06b455aae1b7805f1de389f6868ec7e2887efd6

**Comment by broccoliSpicy** [comment]
in cases like `ms marco`, the third column is way larger(more than 100x) than the other two column, it can be a problem if `do_flush` only produces one encoding task

**Review by westonpace — APPROVED** [review]
Thank you @broccoliSpicy !  I think this PR is in a good spot.  I've filed https://github.com/lancedb/lance/issues/2602 for follow-up tasks.  I want to do just a few more personal tests tomorrow and then I will merge this in.

**Comment by broccoliSpicy** [comment]
@westonpace thank you so much for the review!

I checked out #2602  and we can see how to tackle under #2602

**Comment by westonpace** [comment]
@broccoliSpicy I picked a random parquet file from https://huggingface.co/datasets/HuggingFaceFW/fineweb/tree/main/data

Then I converted the data with (I ran this script twice, once with FSST and once without, and saved the files to two different paths):

```
from lance.file import LanceFileReader, LanceFileWriter
import pyarrow.parquet as pq

data = pq.read_table("data.parquet")
with LanceFileWriter("/home/pace/dev/lance-experiments/fineweb/output.lance") as writer:
  writer.write_batch(data)
```

Then I queried it with:

```
import datetime
import pyarrow.parquet as pq
from lance.file import LanceFileReader

start = datetime.datetime.now()
tab = pq.read_table("data.parquet")
end = datetime.datetime.now()
elapsed = (end - start).total_seconds()
print(f"Parquet elapsed: {elapsed}s")

start = datetime.datetime.now()
tab2 = LanceFileReader("/home/pace/dev/lance-experiments/fineweb/data.lance").read_all().to_table()
end = datetime.datetime.now()
elapsed = (end - start).total_seconds()
print(f"Lv2(no-fsst) elapsed: {elapsed}s")

start = datetime.datetime.now()
tab2 = LanceFileReader("/home/pace/dev/lance-experiments/fineweb/data_fsst.lance").read_all().to_table()
end = datetime.datetime.now()
elapsed = (end - start).total_seconds()
print(f"Lv2(fsst) elapsed: {elapsed}s")

assert tab == tab2

print("Tables are equal")
```

**Comment by broccoliSpicy** [comment]
@westonpace thanks! I will look into this tomorrow. the fsst decoder can be trivial accelerated to 1 byte per 2 - 3 CPU cycle, and I can try put the original size in symbol table.

**Pull request merged**
